### PR TITLE
feat: allow port option when run with target chromium

### PR DESCRIPTION
**Pull request by dever23b**
This suggestion allows us to create a dependable debug workflow to attach an IDE to the browser that gets spawned with the `run` command.

By default, the underlying `chrome-launcher` flow will choose a random debugging port when launching the browser.  This is not ideal for those of us hoping to attach  our IDE/other debugging tool to the browser.  Ideally, we launch the browser with a predictable port that we can configure into our debugging flow.  The underlying `chrome-launcher` actually already supports this; however, `web-ext` does not currently provide any way to pass the required option through.  This PR bridges that gap.

### CODE REVIEW

**Review by Rob--W — CHANGES_REQUESTED** [review]
Someone else attempted the same functionality before. The same review comments that I put there also apply here, please take a look at https://github.com/mozilla/web-ext/pull/2565

**Comment by dever23b** [comment]
Oh, I apologize-- I got ahead of myself and failed to check that other PR.  I see your concern regarding a potential to attach to an instance that doesn't have the extension loaded.  Good thinking!  

I can try to work on a mitigation for that.  Since I haven't gotten started yet, I don't know if what I'm asking for might be more effort than it's worth to write, but are you hard set on aborting if the port is in use?  I'm wondering if it would be permissible to attach to an existing instance on that port, if I could find a way to verify that the existing instance was launched with the same `--load-extension` parameter.

**Comment by dever23b** [comment]
Alright, sorry for the abundance of commits.  I didn't realize that this PR would cause everything to automatically sync from my fork.  

I believe the current state meets your stated concerns, @Rob--W.  Would you mind giving this another review?

File: src/util/verify-chromium-port.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import { createServer } from 'node:http';
```
Comment (Rob--W) [inline src/util/verify-chromium-port.js:1]:
```suggestion
import { createServer } from 'node:net';
```

We don't really need http capabilities, `node:net` is enough and we don't need `node:http`.

File: src/util/verify-chromium-port.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * @param {number} port Port to test
+ * */
+async function isPortAvailable(port) {
+  return new Promise((resolve) => {
+    const server = createServer();
+    server.once('listening', () => {
+      server.close(() => resolve(true));
+    });
+    server.once('error', () => resolve(false));
+    server.listen(port);
```
Comment (Rob--W) [inline src/util/verify-chromium-port.js:16]:
```suggestion
    server.listen(port, '127.0.0.1');
```

By default, `listen()` listens on `::` and/or `0.0.0.0`, which would enable external clients on the network to connect to the server. This may also trigger a firewall notification. This all is undesirable.

To avoid these issues, explicitly pass in the localhost IP, i.e. `127.0.0.1`.

File: src/util/verify-chromium-port.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import { UnusablePortError } from '../errors.js';
+
+/**
+ * Determine if a port is available
+ * @param {number} port Port to test
+ * */
+async function isPortAvailable(port) {
+  return new Promise((resolve) => {
+    const server = createServer();
+    server.once('listening', () => {
```
Comment (Rob--W) [inline src/util/verify-chromium-port.js:12]: For simplicity, you could also consider passing the callback as a parameter to the `listen` function, which is equivalent to manually registering a `listening` listener beforehand.

File: src/extension-runners/chromium.js [hunk 4]
```diff
@@ -151,6 +153,18 @@ export class ChromiumExtensionRunner {
       chromeFlags.push(...this.params.args);
     }
 
+    const { chromiumPort } = this.params;
+    if (chromiumPort) {
```
Comment (Rob--W) [inline src/extension-runners/chromium.js:157]: Are you intentionally excluding `--chromium-port=0` here? The validation helpers that you have added are not 100% correct when `0` is passed.

The expected behavior when `0` is passed is to select a random port (which chrome-launcher does indeed).

File: src/util/verify-chromium-port.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      server.close(() => resolve(true));
+    });
+    server.once('error', () => resolve(false));
```
Comment (Rob--W) [inline src/util/verify-chromium-port.js:15]: This assumes that the only error is that the port is in use (in which case `e.code === 'EADDRINUSE'`). This is not correct. E.g. ports 1-1023 require privileges, and when not granted (which is the most common case), EACCES is raised.

File: src/util/verify-chromium-port.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ * @returns {boolean}
+ */
+function isPortValid(port) {
+  if (!port) {
+    return false;
+  }
+  if (Number.isNaN(port) || !Number.isInteger(port)) {
+    throw new UnusablePortError(`Port provided is not an integer (${port})`);
+  }
+  if (port < 0 || port > 65535) {
```
Comment (Rob--W) [inline src/util/verify-chromium-port.js:32]:
```suggestion
  if (port <= 0 || port > 65535) {
```
`0` should be excluded here. It is not reachable and if it were, it would result in a random port being chosen.

File: src/util/verify-chromium-port.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Validate that requested port is a valid port
+ * @param {number} port Debugging port
+ * @returns {boolean}
+ */
+function isPortValid(port) {
+  if (!port) {
+    return false;
+  }
```
Comment (Rob--W) [inline src/util/verify-chromium-port.js:28]:
```suggestion
```

The caller is expected to pass a port, this branch is unreachable. For simplicity let's just drop it (an error would be raised if `port` is somehow omitted).

File: src/util/verify-chromium-port.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+  return true;
+}
+
+/**
+ * Validate user-supplied port to ensure it is suitable for use
+ * @param {number} port User-supplied port request
+ * @param {string} chromiumBinary Chromium binary being requested for launch
+ * @param {string[]} chromeFlags Array of flags requested for launch
```
Comment (Rob--W) [inline src/util/verify-chromium-port.js:43]:
```suggestion
```

These parameters don't exist.

File: src/util/verify-chromium-port.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    });
+    server.once('error', () => resolve(false));
+    server.listen(port);
+  });
+}
+
+/**
+ * Validate that requested port is a valid port
+ * @param {number} port Debugging port
+ * @returns {boolean}
```
Comment (Rob--W) [inline src/util/verify-chromium-port.js:23]: This `isPortValid` helper` and the exported `isPortValid` helper document a boolean return value, but throw errors when the port is invalid. This is not documented, and strange.

To avoid the strange "true / false / Error` tristate, you could consider dropping the boolean return values and just have error vs non-Error.

```js
// Raises an error if the port is invalid for any reason
await ensureValidPort(port);
```

Another option is to have a boolean return value (port available or not) and throw one error (`PortNotAvailableError`), with the diagnostic information exported via the `log` object ([example of createLogger](https://github.com/mozilla/web-ext/blob/1c015188bc215cdbd327bb5d2213727f178d6231/src/extension-runners/firefox-android.js#L17-L20)).

File: tests/unit/helpers.js [hunk 10]
```diff
@@ -193,6 +193,15 @@ export class StubChildProcess extends EventEmitter {
   stderr = new EventEmitter();
   stdout = new EventEmitter();
   kill = sinon.spy(() => {});
+
+  constructor(params = {}) {
```
Comment (Rob--W) [inline tests/unit/helpers.js:197]: Drop this all - it is not needed by `StubChildProcess` and should not be added.

File: tests/unit/test-extension-runners/test.chromium.js [hunk 11]
```diff
@@ -615,6 +615,36 @@ describe('util/extension-runners/chromium', async () => {
       }),
   );
 
+  it('does pass port to chrome', async () => {
+    const port = 9222;
```
Comment (Rob--W) [inline tests/unit/test-extension-runners/test.chromium.js:619]: This unit test will fail if 9222 is used at any point. You should choose a free port here.

Note that we already have a utility to find a free port, at https://github.com/mozilla/web-ext/blob/1c015188bc215cdbd327bb5d2213727f178d6231/src/firefox/remote.js#L233

File: tests/unit/test-extension-runners/test.chromium.js [hunk 12]
```diff
@@ -23,7 +23,7 @@ import isDirectory from '../../../src/util/is-directory.js';
 
 function prepareExtensionRunnerParams({ params } = {}) {
   const fakeChromeInstance = {
-    process: new StubChildProcess(),
+    process: new StubChildProcess(params),
```
Comment (Rob--W) [inline tests/unit/test-extension-runners/test.chromium.js:26]: If you want to more realistically mimic the result of chromium-launch, then you can assign the result to `fakeChromeInstance` instead of forwarding it through `StubChildProcess`.

Here is the type from chrome-launcher: https://github.com/GoogleChrome/chrome-launcher/blob/v1.1.1/src/chrome-launcher.ts#L85

File: tests/unit/test-extension-runners/test.chromium.js [hunk 13]
```diff
@@ -615,6 +615,36 @@ describe('util/extension-runners/chromium', async () => {
...
+    assert.isDefined(
+      runnerInstance.chromiumInstance,
+      'returned process instance',
+    );
+    assert.isDefined(
+      runnerInstance.chromiumInstance.process.port,
+      'process instance has port value',
+    );
+    assert.equal(
+      runnerInstance.chromiumInstance.process.port,
```
Comment (Rob--W) [inline tests/unit/test-extension-runners/test.chromium.js:640]: `port` is not a member of process. Fix the stub above and read it off `chromiumInstance.port`.

File: tests/unit/test-extension-runners/test.chromium.js [hunk 14]
```diff
@@ -615,6 +615,36 @@ describe('util/extension-runners/chromium', async () => {
...
+      port,
+    });
+    assert.isDefined(
+      runnerInstance.chromiumInstance,
+      'returned process instance',
+    );
+    assert.isDefined(
+      runnerInstance.chromiumInstance.process.port,
+      'process instance has port value',
+    );
```
Comment (Rob--W) [inline tests/unit/test-extension-runners/test.chromium.js:638]:
```suggestion
```

This assertion is not useful. It merely tests the behavior of the internal stub logic.

File: tests/unit/test-extension-runners/test.chromium.js [hunk 15]
```diff
@@ -615,6 +615,36 @@ describe('util/extension-runners/chromium', async () => {
       }),
   );
 
+  it('does pass port to chrome', async () => {
```
Comment (Rob--W) [inline tests/unit/test-extension-runners/test.chromium.js:618]: This unit test only tests the good-weather behavior. You should also test the bad weather behavior, i.e. the error conditions.

Here are the relevant cases:

- port is available.
- port is clearly invalid, e.g. -1 and 99999
- edge case: port 0. Ordinarily this means to select a random port.
- port is unavailable, which the test can guarantee by reserving a port.

**Review by Rob--W — CHANGES_REQUESTED** [review]

**Comment by Rob--W** [comment]
`web-ext` has to switch to communication by pipes to communicate with Chrome to continue support for `web-ext run` (https://github.com/mozilla/web-ext/issues/3388).
That is mutually exclusive with `--remote-debugging-port`, and therefore I am closing this PR.
